internal/consensus/block_executor_test.go

```diff
@@ -0,0 +1,247 @@
+	"github.com/tendermint/tendermint/types/mocks"
+)
+
+type BockExecutorTestSuite struct {
+	suite.Suite
+
+	blockExec     *blockExecutor
+	mockPrivVal   *mocks.PrivValidator
+	mockBlockExec *smmocks.Executor
+}
+
+func TestBockExecutor(t *testing.T) {
+	suite.Run(t, new(BockExecutorTestSuite))
+}
+
+func (suite *BockExecutorTestSuite) SetupTest() {
+	logger := log.NewTestingLogger(suite.T())
+	suite.mockPrivVal = mocks.NewPrivValidator(suite.T())
+	suite.mockBlockExec = smmocks.NewExecutor(suite.T())
+	suite.blockExec = &blockExecutor{
+		logger: logger,
+		privValidator: privValidator{
+			PrivValidator: suite.mockPrivVal,
+			ProTxHash:     crypto.RandProTxHash(),
+		},
+		blockExec:          suite.mockBlockExec,
+		proposedAppVersion: 0,
+	}
+}
+
+func (suite *BockExecutorTestSuite) TestCreate() {
+	ctx := context.Background()
+	commitH99R0 := &types.Commit{
+		Height: 99,
+		Round:  0,
+	}
+	emptyCommit := types.NewCommit(0, 0, types.BlockID{}, nil)
+	testCases := []struct {
+		round         int32
+		initialHeight int64
+		height        int64
+		lastCommit    *types.Commit
+		wantCommit    *types.Commit
+		wantBlock     *types.Block
+		wantCRS       sm.CurrentRoundState
+	}{
+		{
+			round:         0,
+			height:        1,
+			initialHeight: 1,
+			wantCommit:    emptyCommit,
+			wantBlock: &types.Block{
+				Header: types.Header{
+					Height:  1,
+					AppHash: []byte("H1R0"),
+				},
+			},
+			wantCRS: sm.CurrentRoundState{
+				AppHash: []byte("H1R0"),
+			},
+		},
+		{
+			round:         0,
+			height:        100,
+			initialHeight: 1,
+			lastCommit:    commitH99R0,
+			wantCommit:    commitH99R0,
+			wantBlock: &types.Block{
+				Header: types.Header{
+					Height:  1,
+					AppHash: []byte("H100R0"),
+				},
+			},
+			wantCRS: sm.CurrentRoundState{
+				AppHash: []byte("H100R0"),
+			},
+		},
+		{
+			round:         0,
+			height:        100,
+			initialHeight: 1,
+			lastCommit:    nil,
+			wantCommit:    nil,
+		},
+	}
+	for i, tc := range testCases {
+		suite.Run(fmt.Sprintf("test-case #%d", i), func() {
+			stateData := StateData{
+				RoundState: cstypes.RoundState{
+					Height:     tc.height,
+					LastCommit: tc.lastCommit,
+				},
+				state: sm.State{
+					InitialHeight: tc.initialHeight,
+				},
+			}
+			if tc.wantCommit != nil {
+				suite.mockBlockExec.
+					On(
+						"CreateProposalBlock",
+						mock.AnythingOfType("*context.emptyCtx"),
+						tc.height,
+						tc.round,
+						stateData.state,
+						tc.wantCommit,
+						suite.blockExec.privValidator.ProTxHash.Bytes(),
+						suite.blockExec.proposedAppVersion,
+					).
+					Once().
+					Return(tc.wantBlock, tc.wantCRS, nil)
+			}
+			actualBlock, err := suite.blockExec.create(ctx, &stateData, tc.round)
+			require.NoError(suite.T(), err)
+			require.Equal(suite.T(), tc.wantBlock, actualBlock)
+			require.Equal(suite.T(), tc.wantCRS, stateData.CurrentRoundState)
+		})
+	}
+}
+
+func (suite *BockExecutorTestSuite) TestProcess() {
```

Comment:
I'd rather to use suite-test because it helps to separate the logic between initialization and implementation.
`suite` provides several hooks, that can be very useful. For example, the `SetupTest` hook is always called before the test is run, while the `TearDownTest` hook is called after the test is run.
Thus, the benefit of this approach is following
1. you can defines the test dependencies at suite structure and you don't need to create global variables for this
2. initialize deps in `SetupTest`, that will recreate with initial state. so, you never share the state between the tests if you didn't do it deliberately